### PR TITLE
Email Login

### PR DESCRIPTION
**Pull request by soot0-JoshR**
Added functionality for login via email.

### CODE REVIEW

File: src/test/java/dev/devature/penguin_api/service/RegisterServiceTest.java [hunk 1]
```diff
@@ -1,57 +1,54 @@
...
         String[] password = {"1234", "Test_Password12", "TestPassword12", "Test_Password"};
         boolean[] expectedResult = {false, true, false, false};
         boolean[] actualResult = new boolean[4];
 
-        // When
-        for(int i = 0; i < password.length; ++i){
-            actualResult[i] = registerService.checkPasswordRequirements(password[i]);
+        // Check password validity using the utility class
+        for (int i = 0; i < password.length; ++i) {
+            actualResult[i] = ValidationUtils.isValidPassword(password[i]);
```
Comment (eggrain) [inline src/test/java/dev/devature/penguin_api/service/RegisterServiceTest.java]: With the refactor moving those methods to `ValidationUtils`, these are really tests of that class, not `RegisterService` so could you make a test class for that and move these tests there

File: src/main/java/dev/devature/penguin_api/utils/ValidationUtils.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package dev.devature.penguin_api.utils;
+
+import java.util.regex.Pattern;
+
+public class ValidationUtils {
```
Comment (eggrain) [inline src/main/java/dev/devature/penguin_api/utils/ValidationUtils.java]: Maybe the name of this class can be more specific

**Review by eggrain — CHANGES_REQUESTED** [review]
My main concern about this PR is to may be over designing based on some assumptions

I would ask for the PR to be focused on just user login with locally stored email and password. As I am not sure of the exact details of how login will work with the alternative ways, I question if it makes sense for there to be a single endpoint and DTO for our app for these; if it does make sense to have a single endpoint, I think the current implementation can be improved (I am not keen on the controller having to do this stuff where it maps the strategy names, and finds the right strategy to dispatch). One alternative would be for the service that takes the Dto argument to use a switch statement to dispatch based on that `authType`, which also could be an enum type instead of a string, which would avoid the controller needing to know about the strategies. With that the `EmailAuthService` can just be `LoginService` too.

The `AuthenticationController` can just be `LoginController` if you ask me. Authentication is kind of something that happens every request-response (with the json web token in our architecture) - it is identifying the user agent. Logging in the user is definitely related but, in my opinion, `LoginController` is a better abstraction for the controller (to be honest we could refactor to have `UserController` for registering, logging in, logging out, etc but for now I do think this controller can be LoginController)

I would need to look into what the spring-boot-starter-jdbc adds compared to spring-boot-starter-data-jpa. Not sure what you are going for with the changes adding some db configuration for a test environment. Unit tests mock the repository layer, so there is no need for a real database for running these tests against (we can do integration tests when we are further along, I think there are some discussions about the approach to the database to have first)

**Comment by soot0-JoshR** [comment]
@KyleRego, my thoughts:

1) "I would ask for the PR to be focused on just user login with locally stored email and password. As I am not sure of the exact details of how login will work with the alternative ways, I question if it makes sense for there to be a single endpoint and DTO for our app for these; if it does make sense to have a single endpoint"

[The user story for User login](https://github.com/Devature/penguin-api/issues/4) says the user should be able to log in with username and password at /user/login
[The user story for SSO integration](https://github.com/Devature/penguin-api/issues/8) doesn't make a distinction about endpoint. I just designed AuthenticationController around the idea of using a single controller for all log ins as no distinction has been made and it hasn't been discussed previously (to my knowledge).


2) "I think the current implementation can be improved (I am not keen on the controller having to do this stuff where it maps the strategy names, and finds the right strategy to dispatch). One alternative would be for the service that takes the Dto argument to use a switch statement to dispatch based on that authType, which also could be an enum type instead of a string, which would avoid the controller needing to know about the strategies."

If we use different controllers and endpoints for each login method, this is a moot point as there won't be a cental AuthenticationController in which to use a switch-case or enum.


3) "With that the EmailAuthService can just be LoginService too."

Integrating SSO presents the issue where LoginService would then be inaptly named. If EmailAuthService is renamed, it should perhaps be EmailLoginService instead.


4) "The AuthenticationController can just be LoginController if you ask me. Authentication is kind of something that happens every request-response (with the json web token in our architecture) - it is identifying the user agent."

Agreed, AuthenticationController seems odd now that you mention it.


5) "Logging in the user is definitely related but, in my opinion, LoginController is a better abstraction for the controller (to be honest we could refactor to have UserController for registering, logging in, logging out, etc but for now I do think this controller can be LoginController)"

I think this undermines your initial point about focusing on email/password login. I think the controllers and services should be named according to the login method used, especially if we are going to use different endpoints and have this kind of separation of interests in those layers. I don't think we should have a single UserController, or I don't think we should focus on that in this PR at least.


6) "I would need to look into what the spring-boot-starter-jdbc adds compared to spring-boot-starter-data-jpa."

JDBC adds nothing. I didn't realize we were using JPA and haven't removed the dependency yet.


7) "Not sure what you are going for with the changes adding some db configuration for a test environment. Unit tests mock the repository layer, so there is no need for a real database for running these tests against (we can do integration tests when we are further along, I think there are some discussions about the approach to the database to have first)"

I didn't realize the unit tests would mock the repository layer. Either way, it would need to be changed later down the road whether it is removed now or not.

**Pull request closed without merging**